### PR TITLE
{cc,bintools}-wrapper, ghc, libgcc: Define wrapper env vars as full paths

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

This makes builds more robust by having the env vars not rely on the PATH. That, for example, helps keep separate toolchains untangled even when on native builds neither are prefixed.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

CC @bgamari we talked about this once

### CODE REVIEW

File: pkgs/development/compilers/ghc/8.2.2.nix [hunk 1]
```diff
@@ -134,16 +134,16 @@ stdenv.mkDerivation rec {
     done
     # GHC is a bit confused on its cross terminology, as these would normally be
     # the *host* tools.
-    export CC="${targetCC}/bin/${targetCC.targetPrefix}cc"
-    export CXX="${targetCC}/bin/${targetCC.targetPrefix}cxx"
+    export CC="$CC_FOR_TARGET"
+    export CXX="$CXX_FOR_TARGET"
     # Use gold to work around https://sourceware.org/bugzilla/show_bug.cgi?id=16177
     export LD="${targetCC.bintools}/bin/${targetCC.bintools.targetPrefix}ld${stdenv.lib.optionalString targetPlatform.isAarch32 ".gold"}"
```
Comment (matthewbauer) [inline pkgs/development/compilers/ghc/8.2.2.nix]: Why is this one not changed?

Comment (Ericson2314) [reply pkgs/development/compilers/ghc/8.2.2.nix]: Because we are forcing gold to be used

**Comment by Ericson2314** [comment]
@GrahamcOfBorg build stdenv

**Comment by Ericson2314** [comment]
I cannot test on Darwin at the moment, but it seems unlikely that this would cause any problems.

**Comment by vcunat** [comment]
I see a disadvantage: some packages tend to save these variables somewhere in the outputs, so this PR blows runtime closures for those.  Example where it results into a failure due to closure check: https://hydra.nixos.org/build/79797093

Most of these cases probably can be handled, e.g. put those files into `$dev` or something, but as usual, it's takes some work (even to discover the change).

**Comment by Ericson2314** [comment]
Ah, nothing comes free. Thank you very much for catching, @vcunat.

**Comment by vcunat** [comment]
Several packages started failing with some libtool error that _might_ be due to this &ndash; any ideas?  https://hydra.nixos.org/build/79808675

**Comment by vcunat** [comment]
Oh, never mind, it's all the same workaround.  I should have tried before asking :-)

**Comment by vcunat** [comment]
No, my bad, I don't know a workaround.  (I confused two packages together.)

**Comment by matthewbauer** [comment]
This appears to break windows builds too:

https://hydra.nixos.org/build/79957713/

Error is:

```
sh: gcc: not found
```

**Comment by matthewbauer** [comment]
Another case where this breaks stuff:

https://hydra.nixos.org/build/79982541/

To be fair it's kind of dumb what cntlm is doing, but this pr did break it.

**Pull request merged**